flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaRecordSerializationSchemaBuilder.java

```diff
@@ -0,0 +1,270 @@
+ * {@code
+ * new KafkaRecordSerializationSchemaBuilder<String>("topic")
+ *     .withPartitioner(MY_FLINK_PARTITIONER)
+ *     .withKeySerializationSchema(StringSerializer.class)
+ *     .build()
+ * }</pre>
+ *
+ * <p>The different serialization methods for key and value are mutually exclusive thus i.e. it is
+ * not possible to use {@link #withKeySerializationSchema(SerializationSchema)} and {@link
+ * #withKeySerializer(Class)} on the same builder instance.
+ *
+ * <p>It is necessary to configure exactly one serialization method for the value.
+ *
+ * @param <IN> type of records to be serialized
+ * @see KafkaRecordSerializationSchema#builder(String)
+ */
+public class KafkaRecordSerializationSchemaBuilder<IN> {
+
+    @Nullable private final String topic;
+    @Nullable private final Function<IN, String> topicSelector;
+    @Nullable private FlinkKafkaPartitioner<IN> partitioner;
+    @Nullable private SerializationSchema<IN> keySerializationSchema;
+
+    private SerializationSchema<IN> valueSerializationSchema;
+
+    KafkaRecordSerializationSchemaBuilder(String topic) {
+        this.topic = checkNotNull(topic);
+        this.topicSelector = null;
+    }
+
+    KafkaRecordSerializationSchemaBuilder(Function<IN, String> topicSupplier) {
+        this.topicSelector = checkNotNull(topicSupplier);
+        this.topic = null;
+    }
+
+    /**
+     * Sets a custom partitioner determining the target partition of the target topic.
+     *
+     * @param partitioner
+     * @return {@link KafkaRecordSerializationSchemaBuilder}
+     */
+    public KafkaRecordSerializationSchemaBuilder<IN> withPartitioner(
+            FlinkKafkaPartitioner<IN> partitioner) {
+        this.partitioner = checkNotNull(partitioner);
+        return this;
+    }
+
+    /**
+     * Sets a {@link SerializationSchema} which is used to serialize the incoming element to the key
+     * of the {@link ProducerRecord}.
+     *
+     * @param keySerializationSchema
+     * @return {@link KafkaRecordSerializationSchemaBuilder}
+     */
+    public KafkaRecordSerializationSchemaBuilder<IN> withKeySerializationSchema(
+            SerializationSchema<IN> keySerializationSchema) {
+        checkState(this.keySerializationSchema == null, exceedsKeySerializerConfigurations());
+        this.keySerializationSchema = checkNotNull(keySerializationSchema);
+        return this;
+    }
+
+    /**
+     * Sets Kafka's {@link Serializer} to serialize incoming elements to the key of the {@link
+     * ProducerRecord}.
+     *
+     * @param keySerializer
+     * @return {@link KafkaRecordSerializationSchemaBuilder}
+     */
+    public KafkaRecordSerializationSchemaBuilder<IN> withKeySerializer(
+            Class<? extends Serializer<IN>> keySerializer) {
+        checkState(keySerializationSchema == null, exceedsKeySerializerConfigurations());
+        this.keySerializationSchema =
+                new KafkaSerializerWrapper<>(keySerializer, topic, topicSelector);
```

Comment:
Good solution with immediately wrapping things into `KafkaSerializerWrapper`. However, now `topicSelector` is invoked up to 3 times per record. 
There are two solutions: 
- add another indirections `KafkaRecordSerializationSchemaPart` (yes, please find a better name) and let both `KafkaSerializerWrapper` and another class `DeserializationSchemaWrapper` that also take the topic.
- use a stateful topic selector that caches the values for `KafkaSerializerWrapper` (maybe `TopicContext` or so).
After some thinking: I like the second option better.